### PR TITLE
Adjust Hellinger pairwise distance to vaoid NaNs

### PR DESCRIPTION
**Pull request by lowener**
This change will fix NaNs that arise in Hellinger pairwise distance when the input to sqrt is negative.
These kind of inputs can happen even when the inputs are normalized row-wise to 1 due to accumulation of rounding approximation.

### CODE REVIEW

File: cpp/include/raft/sparse/distance/l2_distance.cuh [hunk 1]
```diff
@@ -270,7 +270,11 @@ class hellinger_expanded_distances_t : public distances_t<value_t> {
 
     raft::linalg::unaryOp<value_t>(
       out_dists, out_dists, config_->a_nrows * config_->b_nrows,
-      [=] __device__(value_t input) { return sqrt(1 - input); },
+      [=] __device__(value_t input) {
+        // Adjust to replace NaN in sqrt with 0 if input to sqrt is negative
+        value_t rectifier = (1 - input) > 0;
```
Comment (divyegala) [inline cpp/include/raft/sparse/distance/l2_distance.cuh:275]: This should be bool. Just combine this with next line?

**Review by divyegala — APPROVED** [review]
Pre-approving since it's a small requested change

**Comment by cjnolet** [comment]
We should just run an L1 norm on the input here to guarantee it's positive.  (that's what UMAP does, which provides this with their API).

**Review by cjnolet — CHANGES_REQUESTED** [review]
These are supposed to be probabilities but we should just run an L1 normalization on the input here to guarantee it's positive. (that's what UMAP does, which provides this with their API).

**Comment by lowener** [comment]
Even when the inputs are L1 normalized this can arise. Running `csr_row_normalize_l1` at the start of this function to enforce it doesn't solve these negative values in sqrt. These negative values are often negative epsilon that comes from either the first sqrt or the inner product.

A solution could be to run csr_row_normalize_l1 and keep this rectifier for negative epsilon, or to add an `abs` function call before sqrt.

**Comment by cjnolet** [comment]
I looked at your failed Pytest asserion in cuml. It looks like Hellinger is running into a similar problem that other expanded metrics had been facing where the diagonal (which should be 0's) has a nan. This is likely because of the subtle numerical instabilities w/ expanded euclidean-based metrics (catastrophic cancellations). 

One thing we can do is zero out the diagonal of the resulting matrix using the `initializeDiagonalMatrix` primitive.

**Review by cjnolet — APPROVED** [review]
LGTM

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**